File: src/main/java/gargoyle/ct/pref/impl/CTPreferencesImpl.java
package gargoyle.ct.pref.impl;

import gargoyle.ct.config.CTConfig;
import gargoyle.ct.config.convert.CTUnitConverter;
import gargoyle.ct.config.convert.impl.CTConfigConverter;
import gargoyle.ct.convert.Converter;
import gargoyle.ct.pref.CTPreferences;
import gargoyle.ct.pref.impl.prop.*;
import org.jetbrains.annotations.NotNull;

import java.util.concurrent.TimeUnit;

public class CTPreferencesImpl extends CTBasePreferences implements CTPreferences {
    @SuppressWarnings("MagicNumber")
    public CTPreferencesImpl(@NotNull Class<?> clazz) {
        super(clazz);
        addProperty(new CTPrefBooleanProperty(this, PREF_BLOCK, false));
        CTPrefProperty<Integer> transparencyLevel = new CTPrefIntegerProperty(this, PREF_TRANSPARENCY_LEVEL, 30);
        addProperty(transparencyLevel);
        transparencyLevel.addPropertyChangeListener(event -> {
            if (event.getNewValue() < 1) event.getProperty().set(1);
        });
        addProperty(new CTPrefEnumProperty<>(this, ICON_STYLE.class, PREF_ICON_STYLE, ICON_STYLE.BW));
        addProperty(new CTPrefBooleanProperty(this, PREF_SOUND, true));
        addProperty(new CTPrefBooleanProperty(this, PREF_TRANSPARENCY, true));
        addProperty(new CTPrefEnumProperty<>(this, SUPPORTED_LOCALES.class, SUPPORTED_LOCALES.findSimilar()));
        addProperty(new CTPrefObjectProperty<>(CTConfig.class, this, PREF_CONFIG, new ConfigConverterAdapter()));
    }

    @NotNull
    @Override
    public CTPrefProperty<Boolean> block() {
        return getProperty(PREF_BLOCK);
    }

    @NotNull
    @Override
    public CTPrefProperty<CTConfig> config() {
        return getProperty(PREF_CONFIG);
    }

    @NotNull
    @Override
    public CTPrefProperty<ICON_STYLE> iconStyle() {
        return getProperty(PREF_ICON_STYLE);
    }

    @NotNull
    @Override
    public CTPrefProperty<SUPPORTED_LOCALES> supportedLocales() {
        return getProperty(SUPPORTED_LOCALES.class);
    }

    @NotNull
    @Override
    public CTPrefProperty<Boolean> sound() {
        return getProperty(PREF_SOUND);
    }

    @NotNull
    @Override
    public CTPrefProperty<Boolean> transparency() {
        return getProperty(PREF_TRANSPARENCY);
    }

    @NotNull
    @Override
    public CTPrefProperty<Integer> transparencyLevel() {
        return getProperty(PREF_TRANSPARENCY_LEVEL);
    }

    private static class ConfigConverterAdapter implements Converter<CTConfig> {
        private final CTUnitConverter<CTConfig> converter = new CTConfigConverter();

        @NotNull
        @Override
        public String format(CTConfig data) {
            return converter.format(TimeUnit.MINUTES, data);
        }

        @NotNull
        @Override
        public CTConfig parse(String data) {
            return converter.parse(data);
        }
    }
}

File: src/main/kotlin/gargoyle/ct/pref/impl/CTPreferencesImpl.kt
package gargoyle.ct.pref.impl

import gargoyle.ct.config.CTConfig
import gargoyle.ct.config.convert.CTUnitConverter
import gargoyle.ct.config.convert.impl.CTConfigConverter
import gargoyle.ct.convert.Converter
import gargoyle.ct.pref.CTPreferences
import gargoyle.ct.pref.CTPreferences.IconStyle
import gargoyle.ct.pref.CTPreferences.SupportedLocales
import gargoyle.ct.pref.CTPreferences.SupportedLocales.Companion.findSimilar
import gargoyle.ct.pref.CTPropertyChangeEvent
import gargoyle.ct.pref.impl.prop.CTPrefBooleanProperty
import gargoyle.ct.pref.impl.prop.CTPrefEnumProperty
import gargoyle.ct.pref.impl.prop.CTPrefIntegerProperty
import gargoyle.ct.pref.impl.prop.CTPrefObjectProperty
import gargoyle.ct.pref.impl.prop.CTPrefProperty
import java.util.concurrent.TimeUnit
import kotlin.reflect.KClass

class CTPreferencesImpl(clazz: KClass<*>) : CTBasePreferences(clazz), CTPreferences {
    init {
        addProperty(CTPrefBooleanProperty(this, CTPreferences.PREF_BLOCK, false))
        val transparencyLevel: CTPrefProperty<Int> =
            CTPrefIntegerProperty(this, CTPreferences.PREF_TRANSPARENCY_LEVEL, 30)
        addProperty(transparencyLevel)
        transparencyLevel.addPropertyChangeListener { event: CTPropertyChangeEvent<Int> ->
            if (event.newValue < 1) event.property.set(1)
        }
        addProperty(
            CTPrefEnumProperty(
                this,
                IconStyle::class,
                CTPreferences.PREF_ICON_STYLE,
                IconStyle.BW
            )
        )
        addProperty(CTPrefBooleanProperty(this, CTPreferences.PREF_SOUND, true))
        addProperty(CTPrefBooleanProperty(this, CTPreferences.PREF_TRANSPARENCY, true))
        addProperty(CTPrefEnumProperty(this, SupportedLocales::class, findSimilar()))
        addProperty(
            CTPrefObjectProperty(
                CTConfig::class,
                this,
                CTPreferences.PREF_CONFIG,
                ConfigConverterAdapter()
            )
        )
    }

    override fun block(): CTPrefProperty<Boolean> {
        return getProperty(CTPreferences.PREF_BLOCK)
    }

    override fun config(): CTPrefProperty<CTConfig> {
        return getProperty(CTPreferences.PREF_CONFIG)
    }

    override fun iconStyle(): CTPrefProperty<IconStyle> {
        return getProperty(CTPreferences.PREF_ICON_STYLE)
    }

    override fun supportedLocales(): CTPrefProperty<SupportedLocales> {
        return getProperty(SupportedLocales::class)
    }

    override fun sound(): CTPrefProperty<Boolean> {
        return getProperty(CTPreferences.PREF_SOUND)
    }

    override fun transparency(): CTPrefProperty<Boolean> {
        return getProperty(CTPreferences.PREF_TRANSPARENCY)
    }

    override fun transparencyLevel(): CTPrefProperty<Int> {
        return getProperty(CTPreferences.PREF_TRANSPARENCY_LEVEL)
    }

    private class ConfigConverterAdapter : Converter<CTConfig> {
        private val converter: CTUnitConverter<CTConfig> = CTConfigConverter()
        override fun format(data: CTConfig): String {
            return converter.format(TimeUnit.MINUTES, data)
        }

        override fun parse(data: String): CTConfig {
            return converter.parse(data)
        }
    }
}
